### PR TITLE
Add FusedMultiTransformer fuse pass for GPT3

### PR DESCRIPTION
**Pull request by heavengate**
### PR types
New features

### PR changes
Others

### Describe
**Add FusedMultiTransformer fuse pass for GPT3**
- add passes: `fused_multi_transformer_encoder` `fused_multi_transformer_decoder` `fused_multi_transformer_encoder_fuse_qkv` `fused_multi_transformer_decoder_fuse_qkv`
 
    - fused_multi_transformer_encoder
匹配GPT encoder部分的Transformer Layer(包含multt-head attention和feed forward部分)

转化为`FusedMultiTransformer`融合OP，如图中红框部分即为一层，用于推理加速
infoflow 2022-10-20 19-51-42

    - fused_multi_transformer_decoder
decoder结构与encoder类似，多了`Cache KV`的处理，如图中红框部分
infoflow 2022-10-20 19-54-05
转换为带`TimeStep`输入的`FusedMultiTransformer`融合OP，用于解码
infoflow 2022-10-20 19-55-33

    - fused_multi_transformer_encoder/decoder_fuse_qkv
`fused_multi_transformer_encoder/decoder_fuse_qkv`结构与`fused_multi_transformer_encoder/decoder`类似，差别为Multi-Head Attention例QKV是concat成一个Tensor计算的，计算完后通过split分开，如图中红框所示
infoflow 2022-10-20 19-58-29

- add multi-devices passes: `multi_devices_fused_multi_transformer_encoder_fuse_qkv` `multi_devices_fused_multi_transformer_decoder_fuse_qkv`

    - multi-devices passes
multi-devices passes是在上述pass的基础上插入了`c_identity`和`c_allreduce_sum`卡间通信OP
infoflow 2022-10-20 20-00-49

- add subgraph pass support for the 6 passes
- add `IR_NODE_UNLINK` for unlinking nodes in Graph
- fix `Var->persistent` check segment fault when Variable is null

**TODO**
- 抽取可复用模块，精简代码

### CODE REVIEW

File: paddle/fluid/framework/ir/pass.cc [hunk 1]
```diff
@@ -65,11 +68,34 @@ Graph *Pass::Apply(Graph *graph) const {
       true,
       platform::errors::InvalidArgument(
           "The VarDescs of persistable variable are not consistency."));
-  applied_ = true;
   if (!graph->Has(kPassRecorder)) {
     graph->Set<PassRecorder>(kPassRecorder, new PassRecorder);
   }
   graph->Get<PassRecorder>(kPassRecorder).insert(Type());
+
+  if(graph->IsMainGraph() and "graph_viz_pass"!=Type() and "graph_to_program_pass"!=Type()) {
```
Comment (b3602sss) [inline paddle/fluid/framework/ir/pass.cc]: and?

File: paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.cc [hunk 2]
```diff
@@ -0,0 +1,1726 @@
...
+    IR_NODE_LINK_TO(slice_op, slice_out);
+    IR_NODE_LINK_TO(slice_out, fused_multi_transformer)
+
+    IR_NODE_LINK_TO(fused_multi_transformer, ffn_output);
+  };
+
+  int fusion_count{0};
+  auto handler = [&](const GraphPatternDetector::subgraph_t& subgraph,
+                     Graph* g) {
+    GET_IR_NODE_FROM_SUBGRAPH(input0, input0, fused_multi_transformer_pattern);
```
Comment (b3602sss) [inline paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.cc]: 需要主动调用IsCompat(subgraph, graph)， AddOpCompat才能生效

File: paddle/fluid/framework/ir/pass.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/framework/ir/pass.cc]: 此处判断当前图是MainGraph并且是支持对子图做pass才会走到子图pass匹配

File: paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.cc]: Done, Thanks!

File: paddle/fluid/framework/ir/pass.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/framework/ir/pass.cc]: Done, Thanks!

File: paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.h [hunk 3]
```diff
@@ -0,0 +1,416 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.h]: 2022

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.h (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/framework/ir/fused_multi_transformer_decoder_pass.h]: Done，thanks

**Review by qingqing01 — DISMISSED** [review]

**Review by LielinJiang — DISMISSED** [review]

**Review by qingqing01 — APPROVED** [review]

**Review by LielinJiang — APPROVED** [review]

**Review by raindrops2sea — COMMENTED** [review]
增补一下新加几个pass功能和效果的更具体描述。

**Review by raindrops2sea — APPROVED** [review]

**Review by raindrops2sea — COMMENTED** [review]
训练侧也看一下静态训练的复用问题，@sneaxiy @Xreki @gongweibao

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**